pennylane/plugins/tests/test_gates.py

```diff
@@ -0,0 +1,489 @@
+        [0, 0, np.exp(-1j * theta / 2), 0],
+        [0, 0, 0, np.exp(1j * theta / 2)],
+    ]
+)
+
+# list of all non-parametrized single-qubit gates,
+# along with the PennyLane operation name
+single_qubit = [
+    (qml.PauliX, X),
+    (qml.PauliY, Y),
+    (qml.PauliZ, Z),
+    (qml.Hadamard, H),
+    (qml.S, S),
+    (qml.T, T),
+]
+
+# list of all parametrized single-qubit gates
+single_qubit_param = [
+    # (qml.PhaseShift(0, wires=0), phase_shift),
+    (qml.RX, rx),
+    (qml.RY, ry),
+    (qml.RZ, rz),
+]
+# list of all non-parametrized two-qubit gates
+two_qubit = [(qml.CNOT, CNOT), (qml.SWAP, SWAP), (qml.CZ, CZ)]
+# list of all parametrized two-qubit gates
+two_qubit_param = [(qml.CRZ, crz)]
+# list of all three-qubit gates
+three_qubit = [(qml.Toffoli, toffoli), (qml.CSWAP, CSWAP)]
+
+# single qubit unitary matrix
+U = np.array(
+    [
+        [0.83645892 - 0.40533293j, -0.20215326 + 0.30850569j],
+        [-0.23889780 - 0.28101519j, -0.88031770 - 0.29832709j],
+    ]
+)
+
+# two qubit unitary matrix
+U2 = np.array([[0, 1, 1, 1], [1, 0, 1, -1], [1, -1, 0, 1], [1, 1, -1, 0]]) / np.sqrt(3)
+
+# single qubit Hermitian observable
+A = np.array([[1.02789352, 1.61296440 - 0.3498192j], [1.61296440 + 0.3498192j, 1.23920938 + 0j]])
+
+
+# ===============================================================
+
+
+@flaky(max_runs=10)
+class TestGatesQubit:
+    """Test qubit-based devices' probability vector after application of gates.
+    """
+
+    @pytest.mark.parametrize(
+        "basis_state",
+        [
+            np.array([0, 0, 1, 0]),
+            np.array([0, 0, 1, 0]),
+            np.array([1, 0, 1, 0]),
+            np.array([1, 1, 1, 1]),
+        ],
+    )
+    def test_basis_state(self, device, basis_state, tol, skip_if):
+        """Test basis state initialization."""
+        n_wires = 4
+        dev = device(n_wires)
+        capabilities = dev.__class__.capabilities()
+        skip_if(not capabilities["model"] == "qubit")
+
+        @qml.qnode(dev)
+        def circuit():
+            qml.BasisState(basis_state, wires=range(n_wires))
+            return qml.probs(wires=range(n_wires))
+
+        res = circuit()
+
+        expected = np.zeros([2 ** n_wires])
+        expected[np.ravel_multi_index(basis_state, [2] * n_wires)] = 1
+        assert np.allclose(res, expected, atol=tol(dev.analytic))
+
+    def test_qubit_state_vector(self, device, init_state, tol, skip_if):
+        """Test QubitStateVector initialisation."""
+        n_wires = 1
+        dev = device(n_wires)
+        capabilities = dev.__class__.capabilities()
+        skip_if(not capabilities["model"] == "qubit")
+
+        rnd_state = init_state(n_wires)
+
+        @qml.qnode(dev)
+        def circuit():
+            qml.QubitStateVector(rnd_state, wires=range(n_wires))
+            return qml.probs(range(n_wires))
+
+        res = circuit()
+        expected = np.abs(rnd_state) ** 2
+
+        assert np.allclose(res, expected, atol=tol(dev.analytic))
```

Comment:
Can we have the tolerance fixture depend on the `device` fixture, to automatically get the `analytic` attribute value?

Reply:
That would be cool, but I don't see how it works, since the device is only created in the test. One could use `device_kwargs` but it only contains a specific value for `analytic` if this was set from command line (else it is None and will be defined once the device is created). 

In short: it is only defined inside a test whether the test is run analytic or not.